phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -1119,12 +1124,21 @@ private List<List<Scan>> getParallelScans(byte[] startKey, byte[] stopKey) throw
                 int gpsComparedToEndKey = -1;
                 boolean everNotDelayed = false;
                 while (intersectWithGuidePosts && (endKey.length == 0 || (gpsComparedToEndKey=currentGuidePost.compareTo(endKey)) <= 0)) {
-                    Scan newScan = scanRanges.intersectScan(scan, currentKeyBytes, currentGuidePostBytes, keyOffset,
-                        false);
-                    if (newScan != null) {
-                        ScanUtil.setLocalIndexAttributes(newScan, keyOffset,
-                            regionInfo.getStartKey(), regionInfo.getEndKey(),
-                            newScan.getStartRow(), newScan.getStopRow());
+                    List<Scan> newScans =
+                            scanRanges.intersectScan(scan, currentKeyBytes, currentGuidePostBytes,
+                                keyOffset, false, splitPostfix, getTable().getBucketNum());
+                    for (Scan newScan : newScans) {
+                        if (newScan != null) {
+                            ScanUtil.setLocalIndexAttributes(newScan, keyOffset,
+                                regionInfo.getStartKey(), regionInfo.getEndKey(),
+                                newScan.getStartRow(), newScan.getStopRow());
+                            //TODO why are we passing endKey instead of startKey here ?
+                            scans =
+                                    addNewScan(parallelScans, scans, newScan, currentGuidePostBytes,
+                                        false, regionLocation);
```

Comment:
here `currentGuidePostBytes` could be less than `endKey` right?

Reply:
Thanks for pointing this out, @virajjasani .
This was actually a bug in my patch: 
Even if the new code split a scan because of crossing bucket boundaries, we still used the un-split region's key, which meant that we may have grouped different buckets together, which would have broken ordering.

I have removed the endKey parameter, and now use the new scan's end key.
I have also removed some logic that used the current region's start key as the end key, as it was unneccessary, and incorrect in the merged region case.